File: CODE-OPTIMIZATION/esercizio_3/fortran/matrixmul.f90
module precision_module
   integer, parameter :: dp_kind = kind(1.d0)
   integer, parameter :: sp_kind = kind(1.)
   integer, parameter :: my_kind = dp_kind
end module precision_module
   
module timing_module
   use precision_module
   contains
   subroutine timing(t)
         
   real(my_kind), intent(out) :: t
   integer :: time_array(8)
         
   call date_and_time(values=time_array)
   t = 3600.*time_array(5)+60.*time_array(6)+time_array(7)+time_array(8)/1000.
         
   end subroutine timing
end module timing_module

program mm
!
    use precision_module
    use timing_module
       
    implicit none
!
    integer i,j,k ! index
!    real i,j,k ! index
    integer, parameter :: n=1024            ! size of the matrix
!
    real(my_kind) a(1:n,1:n) ! matrix
    real(my_kind) b(1:n,1:n) ! matrix
    real(my_kind) c(1:n,1:n) ! matrix (destination)
    real(my_kind) time1, time2 ! timing
    real(my_kind) :: mflops, bandwidth, sum
!
    mflops = 2*float(n)*float(n)*float(n)/(1000.0_my_kind*1000.0_my_kind)
    bandwidth = 3*my_kind*float(N)*float(N)/(1024*1024)
!
    call timing(time1)
    call random_number(a)
    call random_number(b)
    c = 0._my_kind 
    call timing(time2)
    write(*,*) "--------------------------------------"
    write(*,*) " Matrix-Matrix Multiplication         "
    write(*,*) " rel. 0, naive multiplication         "
    write(*,*) " size =", n                      
    write(*,*) "--------------------------------------"
    write(*,*) "initialization", time2-time1
    write(*,*) a(n/2,n/2),b(n/2,n/2),c(n/2,n/2)
!
! main loop
!
    call timing(time1)
!
    do j = 1,n
       do k = 1,n
          do i = 1,n
             c(i,j) = c(i,j) + a(i,k)*b(k,j)
          enddo
       enddo
    enddo
!
    call timing(time2)
    print*, "--------------------------------------"
    print*, "time for moltiplication", time2-time1
    print*, "Mflops                 ", mflops/(time2-time1)
    print*, 'bandwidth :  ', bandwidth/(time2-time1)
    print*, "--------------------------------------"

! simple check
    sum = 0._my_kind
    do i=1,n
       do j=1,n 
          sum = sum + c(i,j);
       enddo
    enddo
    write(*,*) 'Check -------------> ',sum
    write(*,*) 'Check -------------> ',c(n/2,n/2)
!
end program mm

File: CODE-OPTIMIZATION/esercizio_3/c/matrixmul.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <time.h>

#define nn (1024)

double a[nn][nn];       /** matrix**/
double b[nn][nn];
double c[nn][nn];

int main()
{
  int k, i, j, ii, jj;
  double sum;
  float time1, time2, dub_time;

  printf("===============================\n");
  printf("matrix-matric product (double precision) \n");
  printf("size %d \n",nn );
  printf("===============================\n");

  /* initialize matrix */
  time1 = clock();
  for (j = 0; j < nn; j++)
    {
    for (i = 0; i < nn; i++)
      {
      a[j][i] = ((double)rand())/((double)RAND_MAX);
      b[j][i] = ((double)rand())/((double)RAND_MAX);
      c[j][i] = 0.0L;		
      }
    }
  time2 = clock();
  dub_time = (time2 - time1)/(double) CLOCKS_PER_SEC;
  printf("Time for init --------->  %f \n", dub_time);

  time1 = clock();

/*  c_ij = c_ij + a_ik * b_kj                                  */ 

  for (i = 0; i < nn; i ++)
       for (k = 0; k < nn; k++)
            for (j = 0; j < nn; j++)
                 c[i][j] = c[i][j] + a[i][k]*b[k][j];

  time2 = clock();
  dub_time = (time2 - time1)/(double) CLOCKS_PER_SEC;
  printf("===============================\n");
  printf("Time  -----------------> %f \n", dub_time);
  printf("Mflops ----------------> %f \n", 
          2.0*nn*nn*nn/(1000*1000*dub_time));

  /* a simple check */
  sum = 0.0;
  for (i = 0; i < nn; i++)
     for (j = 0; j < nn; j++)
         sum = sum + c[i][j];
  printf("Check -------------> %f \n", sum);
  printf("Check -------------> %f \n", c[nn/2][nn/2]);

   return 0;  
}
